opm/autodiff/CPRPreconditioner.hpp

```diff
@@ -67,234 +115,124 @@ template<class M, class X, class Y, class P>
+                                    const P& comm)
 {
-    typedef typename CPRSelector<M,X,X,Dune::OwnerOverlapCopyCommunication<I1,I2> >
+    typedef typename CPRSelector<M,X,X,P >
         ::EllipticPreconditioner ParallelPreconditioner;
 
-    //Dune::SeqILU0<M,X,X>* ilu=new Dune::SeqILU0<M,X,X>(Ae, relax);
-    typedef typename CPRSelector<M,X,X,Dune::OwnerOverlapCopyCommunication<I1,I2> >
+    typedef typename CPRSelector<M,X,X,P>
         ::EllipticPreconditionerPointer EllipticPreconditionerPointer;
     return EllipticPreconditionerPointer(new ParallelPreconditioner(Ae, comm, relax));
 }
 
-#endif // #if HAVE_MPI
-
-/// \brief Creates the elliptic preconditioner (ILU0)
-/// \param opA     The operator representing the matrix of the system.
-/// \param relax   The relaxation parameter for ILU0.
-/// \param comm    The object describing the parallelization information and communication.
-//  \param amgPtr  The unique_ptr to be filled (return)
-template < int pressureIndex=0, class Op, class P, class AMG >
+template < class C, class Op, class P, class S, std::size_t index >
 inline void
-createAMGPreconditionerPointer( Op& opA, const double relax, const P& comm, std::unique_ptr< AMG >& amgPtr )
+createAMGPreconditionerPointer(Op& opA, const double relax, const P& comm,
+                               std::unique_ptr< BlackoilAmg<Op,S,C,P,index> >& amgPtr,
+                               const CPRParameter& params = CPRParameter())
 {
-    // type of matrix
-    typedef typename Op::matrix_type  M;
-
-    // The coupling metric used in the AMG
-    typedef Dune::Amg::Diagonal<pressureIndex> CouplingMetric;
+    using AMG = BlackoilAmg<Op,S,C,P,index>;
+    // TODO: revise choice of parameters
+    int coarsenTarget=1200;
+    using Criterion = C;
+    Criterion criterion(15, coarsenTarget);
+    criterion.setDebugLevel( 0 ); // no debug information, 1 for printing hierarchy information
+    criterion.setDefaultValuesIsotropic(2);
+    criterion.setNoPostSmoothSteps( 1 );
+    criterion.setNoPreSmoothSteps( 1 );
 
-    // The coupling criterion used in the AMG
-    typedef Dune::Amg::SymmetricCriterion<M, CouplingMetric> CritBase;
+    // Since DUNE 2.2 we also need to pass the smoother args instead of steps directly
+    typedef typename AMG::Smoother Smoother;
+    typedef typename Dune::Amg::SmootherTraits<Smoother>::Arguments  SmootherArgs;
+    SmootherArgs  smootherArgs;
+    smootherArgs.iterations = 1;
+    smootherArgs.relaxationFactor = relax;
 
-    // The coarsening criterion used in the AMG
-    typedef Dune::Amg::CoarsenCriterion<CritBase> Criterion;
+    amgPtr.reset( new AMG( params, opA, criterion, smootherArgs, comm ) );
+}
 
+template < class C, class Op, class P, class AMG >
+inline void
+createAMGPreconditionerPointer(Op& opA, const double relax, const P& comm, std::unique_ptr< AMG >& amgPtr,
+                               const CPRParameter& params = CPRParameter())
```

Comment:
This parameter is unused. Did you intend to use it to control what is now hardcoded below?